systems/controllers/constrained_lqr_controller.cc

```diff
@@ -0,0 +1,141 @@
+  output_port_efforts_index_ =
+      this->DeclareVectorOutputPort(TimestampedVector<double>(num_efforts_),
+                                    &ConstrainedLQRController::CalcControl)
+          .get_index();
+
+  // checking the validity of the parameters
+  DRAKE_DEMAND(q.size() == num_positions_);
+  DRAKE_DEMAND(u.size() == num_efforts_);
+  DRAKE_DEMAND(lambda.size() == num_forces_);
+
+  // Creating the full state vector (Velocities are zero as it is a fixed point)
+  VectorXd x(num_states_);
+  VectorXd v = VectorXd::Zero(num_velocities_);
+  x << q, v;
+  desired_state_ = x;
+
+  // Computing the full Jacobian (Position and contact)
+  MatrixXd J_tree, J_contact;
+  KinematicsCache<double> k_cache = tree_.doKinematics(q);
+
+  // Position Jacobian
+  J_tree = tree_.positionConstraintsJacobian(k_cache);
+
+  // Contact Jacobian (If contact information is provided)
+  if (contact_info_.num_contacts > 0) {
+    J_contact = autoDiffToValueMatrix(
+        contact_toolkit_->CalcContactJacobian(initializeAutoDiff(x)));
+  }
+
+  MatrixXd J(J_tree.rows() + J_contact.rows(), num_positions_);
+  J << J_tree, J_contact;
+
+  // Computing F
+  MatrixXd F = MatrixXd::Zero(2 * J.rows(), 2 * J.cols());
+  F.block(0, 0, J.rows(), J.cols()) = J;
+  F.block(J.rows(), J.cols(), J.rows(), J.cols()) = J;
+
+  // Computing the null space of F
+  HouseholderQR<MatrixXd> qr_decomp(F.transpose());
+  MatrixXd q_decomp = qr_decomp.householderQ();
+
+  MatrixXd P =
+      q_decomp.block(0, F.rows(), q_decomp.rows(), q_decomp.cols() - F.rows());
+  P.transposeInPlace();
+
+  // To compute the linearization, xdot (autodiff) is computed
+  // Creating a combined autodiff vector and then extracting the individual
+  // components to ensure proper gradient initialization.
+  VectorXd xul(num_states_ + num_efforts_ + num_forces_);
+  xul << x, u, lambda;
+  AutoDiffVecXd xul_autodiff = initializeAutoDiff(xul);
+  AutoDiffVecXd x_autodiff = xul_autodiff.head(num_states_);
+  AutoDiffVecXd u_autodiff = xul_autodiff.segment(num_states_, num_efforts_);
+  AutoDiffVecXd lambda_autodiff = xul_autodiff.tail(num_forces_);
+
+  // xdot
+  AutoDiffVecXd xdot_autodiff = contact_toolkit_->CalcTimeDerivatives(
+      x_autodiff, u_autodiff, lambda_autodiff);
+
+  // Making sure that the derivative is zero
+  DRAKE_DEMAND(autoDiffToValueMatrix(xdot_autodiff).isZero(1e-6));
+
+  MatrixXd AB = autoDiffToGradientMatrix(xdot_autodiff);
+  MatrixXd A = AB.leftCols(num_states_);
+  MatrixXd B = AB.block(0, num_states_, AB.rows(), num_efforts_);
+
+  // A and B matrices in the new coordinates
+  A_ = P * A * P.transpose();
+  B_ = P * B;
+  // Remapping the Q costs to the new coordinates
+  Q_ = P * Q * P.transpose();
+  R_ = R;
+
+  // Validating the dimesions of A_, B_, Q_ and R_ with respect to each other.
+  DRAKE_DEMAND(Q_.rows() == Q_.cols());
+  DRAKE_DEMAND(R_.rows() == R_.cols());
+  DRAKE_DEMAND(A_.rows() == Q_.rows());
+  DRAKE_DEMAND(B_.cols() == R_.rows());
```

Comment:
Aren't all of these checks guaranteed by the linear algebra above? I think the only one that isn't guaranteed is the last `B.cols == R.rows`. It's probably better to check that `Q` and `R` inputs have appropriate size, then assume that the matrix multiplies result in proper dimensions.